pkg/extprom/http/instrument_server.go

```diff
@@ -90,17 +95,72 @@ func NewInstrumentationMiddleware(reg prometheus.Registerer, buckets []float64)
 // value. http_requests_total is a metric vector partitioned by HTTP method
 // (label name "method") and HTTP status code (label name "code").
 func (ins *defaultInstrumentationMiddleware) NewHandler(handlerName string, handler http.Handler) http.HandlerFunc {
-	return promhttp.InstrumentHandlerDuration(
-		ins.requestDuration.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-		promhttp.InstrumentHandlerRequestSize(
-			ins.requestSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-			promhttp.InstrumentHandlerCounter(
-				ins.requestsTotal.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-				promhttp.InstrumentHandlerResponseSize(
-					ins.responseSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
-					handler,
-				),
+	return promhttp.InstrumentHandlerRequestSize(
+		ins.requestSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+		promhttp.InstrumentHandlerCounter(
+			ins.requestsTotal.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+			promhttp.InstrumentHandlerResponseSize(
+				ins.responseSize.MustCurryWith(prometheus.Labels{"handler": handlerName}),
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					now := time.Now()
+
+					wd := &responseWriterDelegator{w: w}
+					handler.ServeHTTP(wd, r)
+
+					observer := ins.requestDuration.WithLabelValues(
+						wd.Status(),
+						handlerName,
+						strings.ToLower(r.Method),
+					)
+					observer.Observe(time.Since(now).Seconds())
+
+					// If we find a tracingID we'll expose it as Exemplar.
+					span := opentracing.SpanFromContext(r.Context())
+					if span != nil {
```

Comment:
Good for now, but maybe we can create some way helpers for this (:

Reply:
Yes, that might be worth doing. 
I also looked into bringing parts upstream into client_golang but almost all of these modifications happen in the http.HandlerFunc which is too specific for client_golang it seemed like.

I can quickly try to refactor into a helper. :+1:

Reply:
Now I have locally:

```go
// within the handler
durationObserver := ins.requestDuration.WithLabelValues(
	wd.Status(),
	handlerName,
	strings.ToLower(r.Method),
)
observeDuration(durationObserver, start, r)

// outside the handler as a helper func

func observeDuration(observer prometheus.Observer, start time.Time, r *http.Request) {
	observer.Observe(time.Since(start).Seconds())

	// If we find a tracingID we'll expose it as Exemplar.
	span := opentracing.SpanFromContext(r.Context())
	if span != nil {
		spanCtx, ok := span.Context().(jaeger.SpanContext)
		if ok {
			observer.(prometheus.ExemplarObserver).ObserveWithExemplar(
				time.Since(start).Seconds(),
				prometheus.Labels{
					"traceID": spanCtx.TraceID().String(),
				},
			)
		}
	}
}
```
WDYT?

Reply:
Yup, that would work, but I think for one example is not worth refactoring 

Once we have 3+ places to use this then it's worth our work, we will also know more! Thanks!